### ISSUE TITLE
Audit file permissions for directories and files created by client and renewer

### ISSUE THREAD

**Issue by schoen** [certbot/certbot#420]
I don't specifically know that they're wrong, but we need to check them to make sure they're all reasonable.

**Comment by schoen** [comment]
@jsha noticed that the client currently wants some directories to have mode 755 and this is mandatory. I'm not sure that's the right policy.

**Comment by pde** [comment]
Results of a partial audit, observing the *results* of running LE rather than reading all relevant code:

Everything under /etc/letsencrypt/ is 700, except for `renewal/`, `csr/`, and options-ssl-apache.conf.  I actually don't understand how webservers that run as something other than root can get things from `live/` and `keys/` right now.

Everything under `/var/log/letsencrypt` is locked down.

There are world-readable backups of some things in `/var/lib/letsencrypt/backups`, but they appear to be config files and not keys.  Since apache does not contain inline private keys or basic auth credentials, those should not be leaked by those permissions.  However `/var/lib/letsencrypt/backups` should be less widely readable if possible.

**Comment by pde** [comment]
Opened #1396

**Comment by pde** [comment]
I think that's sufficient confidence to kick further work on this down the road for now.

**Comment by acrolink** [comment]
So, how to make it possible for a Rails server run as a normal user to access the keys on /live directory?

**Comment by jsha** [comment]
@acrolink: Could you post your question at https://community.letsencrypt.org/? This issue is about a different question. Thanks!

**Comment by sydneyli** [comment]
I did an audit of the files in all the `700` directories in `config-dir` (usually `/etc/letsencrypt/`), where their permissions are set/enforced/created (it's not comprehensive-- permissions could *also* be set elsewhere, but this is intended as a starting point). Links are to lines in commit [`92645619`](https://github.com/certbot/certbot/tree/9264561944fc847e4a5d2de9a2be1c7da811349d).
 * `accounts/<endpoint>/directory/<hash>/`  all [700](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/account.py#L144), enforced
    * meta.json              [default (644)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/account.py#L350)
    * private_key.json       [400](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/account.py#L347)
    * regr.json              [default (644)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/account.py#L330)

* `archive/` and `live/`     [700](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/storage.py#L1022)
   * `<lineage>/`              [default (755)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/storage.py#L1042)
      * fullchain.pem          [default (644)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/storage.py#L1063)
      * cert.pem               [default (644)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/storage.py#L1053)
      * chain.pem              [default (644)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/storage.py#L1060)
      * privkey.pem            [default (644)](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/storage.py#L1056)

 * keys/                  [700](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/crypto_util.py#L64), enforced
    * `####_key-certbot.pem`   [600](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/crypto_util.py#L67)

"enforced" means the permissions are verified with [`util.make_or_verify_dir`](https://github.com/certbot/certbot/blob/9264561944fc847e4a5d2de9a2be1c7da811349d/certbot/util.py#L167). "default" means that the permissions are not explicitly set upon file/directory creation; in parentheses is what these files end up being when created.

As I mentioned in #1473, the biggest problem is that the permissions on `archive`/`live` are not verified or ensured to be 700, so making that directory more permissive in future versions of Certbot means that downgrading Certbot and renewing would expose private key material, since the new private keys would be created under 644.

**Comment by maurorappa** [comment]
This is my server directory, totally managed by cerbot, I never changed any permission:

```
 # ls -l /etc/letsencrypt/archive/dns.0latency.com/
total 48
-rw-r--r-- 1 root root 2159 Aug 17 21:46 cert1.pem
-rw-r--r-- 1 root root 2155 Oct 28 21:48 cert2.pem
-rw-r--r-- 1 root root 1911 Jan 11 15:46 cert3.pem
-rw-r--r-- 1 root root 1647 Aug 17 21:46 chain1.pem
-rw-r--r-- 1 root root 1647 Oct 28 21:48 chain2.pem
-rw-r--r-- 1 root root 1647 Jan 11 15:46 chain3.pem
-rw-r--r-- 1 root root 3806 Aug 17 21:46 fullchain1.pem
-rw-r--r-- 1 root root 3802 Oct 28 21:48 fullchain2.pem
-rw-r--r-- 1 root root 3558 Jan 11 15:46 fullchain3.pem
-rw-r--r-- 1 root root 1704 Aug 17 21:46 privkey1.pem
-rw-r--r-- 1 root root 1704 Oct 28 21:48 privkey2.pem
-rw-r--r-- 1 root root 1704 Jan 11 15:46 privkey3.pem
```

I think it's way too permissing, expecially for the Private key.
I am running certbot version: 0.10.2, on Debian with default umask of 0022.

**Comment by sydneyli** [comment]
Hi @maurorappa! We fixed this in #6480, which should be in the most recent release of Certbot v0.29.1. Regardless, your private key should still be safe as long as the permissions on the containing directories have remained the same (non group or world readable) as when Certbot created them.

**Comment by m0namon** [comment]
It seems that we finished the audit, so closing.

If we want to do a file directory structure rehaul in the future, that should go in a new issue.
